flow/connectors/clickhouse/qrep.go

```diff
@@ -0,0 +1,366 @@
+	`
+	queryString := fmt.Sprintf(schemaStatement, "desti", qRepMetadataTableName) //c.metadataSchema,
+
+	//_, err := createMetadataTableTx.Exec(queryString)
+	_, err := c.database.Exec(queryString)
+	//_, err := c.database.Exec("select * from tasks;")
+
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("failed to create table %s.%s", c.metadataSchema, qRepMetadataTableName),
+			slog.Any("error", err))
+
+		return fmt.Errorf("failed to create table %s.%s: %w", c.metadataSchema, qRepMetadataTableName, err)
+	}
+	//c.logger.Info(fmt.Sprintf("Created table %s", qRepMetadataTableName))
+	return nil
+}
+
+func (c *ClickhouseConnector) createStage(stageName string, config *protos.QRepConfig) error {
+	var createStageStmt string
+	if strings.HasPrefix(config.StagingPath, "s3://") {
+		stmt, err := c.createExternalStage(stageName, config)
+		if err != nil {
+			return err
+		}
+		createStageStmt = stmt
+	} else {
+		stageStatement := `
+			CREATE OR REPLACE STAGE %s
+			FILE_FORMAT = (TYPE = AVRO);
+			`
+		createStageStmt = fmt.Sprintf(stageStatement, stageName)
+	}
+	// Execute the query
+	_, err := c.database.Exec(createStageStmt)
+
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("failed to create stage %s", stageName), slog.Any("error", err))
+		return fmt.Errorf("failed to create stage %s: %w", stageName, err)
+	}
+
+	c.logger.Info(fmt.Sprintf("Created stage %s", stageName))
+	return nil
+}
+
+func (c *ClickhouseConnector) createExternalStage(stageName string, config *protos.QRepConfig) (string, error) {
+	awsCreds, err := utils.GetAWSSecrets(utils.S3PeerCredentials{})
+	if err != nil {
+		c.logger.Error("failed to get AWS secrets", slog.Any("error", err))
+		return "", fmt.Errorf("failed to get AWS secrets: %w", err)
+	}
+
+	s3o, err := utils.NewS3BucketAndPrefix(config.StagingPath)
+	if err != nil {
+		c.logger.Error("failed to extract S3 bucket and prefix", slog.Any("error", err))
+		return "", fmt.Errorf("failed to extract S3 bucket and prefix: %w", err)
+	}
+
+	cleanURL := fmt.Sprintf("s3://%s/%s/%s", s3o.Bucket, s3o.Prefix, config.FlowJobName)
+
+	s3Int := config.DestinationPeer.GetClickhouseConfig().S3Integration
+	if s3Int == "" {
+		credsStr := fmt.Sprintf("CREDENTIALS=(AWS_KEY_ID='%s' AWS_SECRET_KEY='%s')",
+			awsCreds.AccessKeyID, awsCreds.SecretAccessKey)
+
+		stageStatement := `
+		CREATE OR REPLACE STAGE %s
+		URL = '%s'
+		%s
+		FILE_FORMAT = (TYPE = AVRO);`
+		return fmt.Sprintf(stageStatement, stageName, cleanURL, credsStr), nil
+	} else {
+		stageStatement := `
+		CREATE OR REPLACE STAGE %s
+		URL = '%s'
+		STORAGE_INTEGRATION = %s
+		FILE_FORMAT = (TYPE = AVRO);`
+		return fmt.Sprintf(stageStatement, stageName, cleanURL, s3Int), nil
+	}
+}
+
+func (c *ClickhouseConnector) ConsolidateQRepPartitions(config *protos.QRepConfig) error {
+	c.logger.Error("Consolidating partitions")
+
+	destTable := config.DestinationTableIdentifier
+	stageName := c.getStageNameForJob(config.FlowJobName)
+
+	colInfo, err := c.getColsFromTable(destTable)
```

Comment:
this function now returns two string arrays of names/types instead of `map[string]string` in order to maintain column ordering